infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,217 @@
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.protocols.wireprotocol.TokenResponse;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.NetworkException;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.util.CFUtils;
+import org.corfudb.util.LambdaUtils;
+import org.corfudb.util.Sleep;
+import org.corfudb.util.concurrent.SingletonResource;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * An auto-commit service that periodically commits the unwritten addresses
+ * in the global log, continuously consolidating the log prefix.
+ * <p>
+ * Created by WenbinZhu on 5/5/20.
+ */
+@Slf4j
+public class AutoCommitService implements ManagementService {
+
+    private static final int COMMIT_BATCH_SIZE = 500;
+    private static final int MAX_COMMIT_RETRY = 8;
+    private static final Duration CONN_RETRY_RATE = Duration.ofMillis(500);
+
+    private final ServerContext serverContext;
+    private final SingletonResource<CorfuRuntime> runtimeSingletonResource;
+    private final ScheduledExecutorService autoCommitScheduler;
+
+    // Global log tail fetched at last auto commit cycle, which
+    // would be the commit upper bound in the current cycle.
+    private long lastLogTail = Address.NON_ADDRESS;
+    // Cached committed tail so that we do not need to fetch every cycle.
+    private long committedTail = Address.NON_ADDRESS;
+
+    AutoCommitService(@NonNull ServerContext serverContext,
+                      @NonNull SingletonResource<CorfuRuntime> runtimeSingletonResource) {
+        this.serverContext = serverContext;
+        this.runtimeSingletonResource = runtimeSingletonResource;
+        this.autoCommitScheduler = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setDaemon(true)
+                        .setNameFormat(serverContext.getThreadPrefix() + "AutoCommitService")
+                        .build());
+    }
+
+    CorfuRuntime getCorfuRuntime() {
+        return runtimeSingletonResource.get();
+    }
+
+    /**
+     * Starts the long running service.
+     *
+     * @param interval interval to run the service
+     */
+    @Override
+    public void start(Duration interval) {
+        autoCommitScheduler.scheduleAtFixedRate(
+                () -> LambdaUtils.runSansThrow(this::runAutoCommit),
+                interval.toMillis() / 2,
+                interval.toMillis(),
+                TimeUnit.MILLISECONDS
+        );
+    }
+
+    @VisibleForTesting
+    public synchronized void runAutoCommit() {
```

Comment:
The method seems too complicated. Since it is complex, it's hard to write tests for this method. Can you make it smaller and add more unit tests?

![image](https://user-images.githubusercontent.com/314292/84095821-27323d00-a9b5-11ea-8561-5e12a594bb4b.png)

Reply:
It's not complited, the core logic of this method is just 30 lines, others are just exception handling and comments.